structure/set/set.go

```diff
@@ -88,10 +96,18 @@ func (st *set) IsSubsetOf(superSet Set) bool {
 	return true
 }
 
+func (st *set) IsProperSubsetOf(superSet Set) bool {
+	return st.IsSubsetOf(superSet) && st.Len() != superSet.Len()
```

Comment:
I think you should flip the condition. The first thing to check for proper subset is size usually and its more readable. (I think it might even be more efficient as well though readability takes precedence in my mind)

```suggestion
	if st.Len() == superSet.Len() {
		return false
	}
	return st.IsSubsetOf(superSet)
```